pkg/resource/deploy/source_eval.go

```diff
@@ -1008,6 +1021,76 @@ func inheritFromParent(child resource.Goal, parent resource.Goal) *resource.Goal
 	return &goal
 }
 
+type sourcePositions struct {
+	projectRoot string
+}
+
+func newSourcePositions(projectRoot string) *sourcePositions {
+	if projectRoot == "" {
+		projectRoot = "/"
+	} else {
+		contract.Assertf(filepath.IsAbs(projectRoot), "projectRoot is not an absolute path")
+		projectRoot = filepath.Clean(projectRoot)
+	}
+	return &sourcePositions{projectRoot: projectRoot}
+}
+
+// source position URLs:
+// scheme://path#line[,col]
+func (s *sourcePositions) parseSourcePosition(raw string) (string, error) {
+	posURL, err := url.Parse(raw)
+	if err != nil {
+		return "", err
+	}
+	if posURL.Scheme != "file" {
+		return "", fmt.Errorf("unrecognized scheme %q", posURL.Scheme)
+	}
+	line, col, _ := strings.Cut(posURL.Fragment, ",")
+	if line == "" {
+		return "", fmt.Errorf("missing line number")
+	}
+	if _, err := strconv.ParseInt(line, 10, 0); err != nil {
+		return "", fmt.Errorf("invalid line number %q: %w", line, err)
+	}
+	if col != "" {
+		if _, err := strconv.ParseInt(col, 10, 0); err != nil {
+			return "", fmt.Errorf("invalid column number %q: %w", col, err)
+		}
+	}
+
+	file := filepath.FromSlash(posURL.Path)
+	if !filepath.IsAbs(file) {
+		return "", fmt.Errorf("source positions must include absolute paths")
+	}
+	rel, err := filepath.Rel(s.projectRoot, file)
+	if err != nil {
+		return "", fmt.Errorf("making relative path: %w", err)
+	}
+
+	logging.V(5).Infof("(%q, %q) -> %q", file, s.projectRoot, rel)
+
+	posURL.Scheme = "project"
+	posURL.Path = "/" + filepath.ToSlash(rel)
+
+	return posURL.String(), nil
+}
+
+// getFromMetadata returns any source position metadata from an incoming request.
+func (s *sourcePositions) getFromMetadata(ctx context.Context) string {
+	md, _ := metadata.FromIncomingContext(ctx)
+	vals, ok := md["source-position"]
+	if !ok || len(vals) == 0 {
+		return ""
+	}
+
+	pos, err := s.parseSourcePosition(vals[0])
```

Comment:
Should we warn if someone tries to send multiple "source-position" values?

Reply:
I don't think so.

One idea I have for extending this is to support multiple source position values to represent a stack trace, and so I'd rather treat extra source positions as ignored for now.